### DISCUSSION TITLE
Milvus query failing with text having \n or % or any other special characters

### DISCUSSION THREAD

**Discussion by karthikeyan-balu** [milvus-io/milvus#25388] [Q&A and General discussion]
- we are using pymilvus library in python to connect to milvus and perform some operations. so whenever we query some text which is having text with special, milvus is throwing, saying invalid expression for the query. 

how to solve this issue ?

**Comment by yhmo** [comment]
Could you show your query expression?

**Reply by karthikeyan-balu** [reply]
@yhmo 
Here is query expr 

`company_name == "google" and company_text == "Based on Ministry of Corporate Affairs \n The company is defined as standard and independant "`

In vector db, `\n` is inserted as complete link break but unable to query those data basically. 

Getting error as ` error: line 1:47 token recognition error at: `

**Reply by karthikeyan-balu** [reply]
similarly when you query expr contains like 

`expr = company_name like "%name%" ` is not working

**Reply by xiaofan-luan** [reply]
1. milvus don't support anything like %name%
2. try Escape character if you want to match \n. if you write a short code to reproduce the issue it would be really helpful

**Reply by yhmo** [reply]
@karthikeyan-balu
I just verify that even by using Escape character, it cannot match the string that contains "\n" character, neither by "==" or "like". This is the current limitation of hybrid search. You can create an issue for Milvus so that we can fix it as an enhancement.

For "%", assume a collection has a"title" field, the value is: "ABC%CDF". The hybrid search also has bug.
- search(expr="title==\"ABC%CDF\"") can match correctly
- search(expr="title like \"ABC%%\"") can match correctly
- search(expr="title like \"ABC%CD%\"") throw error

**Reply by xiaofan-luan** [reply]
this is not a bug, milvus only support prefix filtering.

**Reply by xiaofan-luan** [reply]
So %A%, or A%B% will not working

**Reply by karthikeyan-balu** [reply]
But how to handle this case, because we used to insert company data which will have text  like  `Based\n\n google's report on \" we handle it \"  gained 10% interest `.  

And also what is the best way to escape special characters in Python on query expr?

**Reply by xiaofan-luan** [reply]
we are trying to improve the string handling to support \n in a graph

**Reply by aladeen57** [reply]
Hello,
I'm currently having the same issue, I'm trying to query with prefix filtering, but also have "%" in my expression. Milvus thinks i'm trying to do prefix and suffix, even with using "\\" before the "%" in the middle of the expression.
When I don't have any "%" in the expression besides the one at the end to do prefix filtering it works perfectly.

Maybe a fix has been done since then, any news ? I'm not able to find any solution on the documentation.

**Reply by xiaofan-luan** [reply]
you need to escape %.
search for %a you get aa,ba,%a
search for \%a you will only get "%a"

**Reply by yhmo** [reply]
@aladeen57 
Could you show your expressions?

**Reply by aladeen57** [reply]
@xiaofan-luan Thanks for the reply, I'm sorry but I didn't get what you meant by escape %. I also don't understand the difference between your two examples, which also seem to be suffixes.

@yhmo Here are my current tries that don't work : 

First exemple of query expression : 
field1== 'XXXX' and field2== 'XXXX' and field3 like "{\\"h2\\": \\"Performance  sur 5 ans +33,9% +29,7% +6,2%  -9,5%  -1,53%\\"%"

associated error : 
pymilvus.decorators - ERROR - RPC error: [query], <MilvusException: (code=65535, message=cannot parse expression: field1== 'XXXX' and field2== 'XXXX' and field3 like "{\\"h2\\": \\"Performance  sur 5 ans +33,9% +29,7% +6,2%  -9,5%  -1,53%\\"%" error: unsupported pattern: {\\"h2\\": \\"Performance  sur 5 ans +33,9% +29,7% +6,2%  -9,5%  -1,53%\\"%, only prefix pattern match like ab% and equal match like ab(no wildcards) are supported)>

second exemple of query expression : 
same as first exemple but with \ or \\ in front my intermediat "%" (in this exemple only one \)
field1== 'XXXX' and field2== 'XXXX' and field3 like "{\\"h2\\": \\"Performance  sur 5 ans +33,9\\% +29,7\\% +6,2\\%  -9,5\\%  -1,53\\%\\"%"

associated error : 
no error but query doesn't work I don't have any match altho I'm sure there should be (I used this query and have results : field1== 'XXXX' and field2== 'XXXX' and field3 like "{\\"h2\\": \\"Performance %", same prefix query but without the intermediate %).

**Reply by xiaofan-luan** [reply]
you have to use a ="99.9%" or "a like "99.9\%""

**Comment by xiaofan-luan** [comment]
% is a wildcard match token, if you do want a real %, you need to add \ as escape before it
